<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<style type="text/css">
@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
}
a {
background-color: transparent;
}
a:active,
a:hover {
outline: 0;
}
strong {
font-weight: bold;
}
h1 {
font-size: 2em;
margin: 0.67em 0;
}
img {
border: 0;
}
hr {
box-sizing: content-box;
height: 0;
}
pre {
overflow: auto;
}
code,
kbd,
pre {
font-family: monospace, monospace;
font-size: 1em;
}
input {
color: inherit;
font: inherit;
margin: 0;
}
html input[disabled] {
cursor: default;
}
input {
line-height: normal;
}
input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
table {
border-collapse: collapse;
border-spacing: 0;
}
td,
th {
padding: 0;
}
* {
box-sizing: border-box;
}
input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
a {
color: #4078c0;
text-decoration: none;
}
a:hover,
a:active {
text-decoration: underline;
}
hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
hr:before {
display: table;
content: "";
}
hr:after {
display: table;
clear: both;
content: "";
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
h1 {
font-size: 30px;
}
h2 {
font-size: 21px;
}
h3 {
font-size: 16px;
}
h4 {
font-size: 14px;
}
h5 {
font-size: 12px;
}
h6 {
font-size: 11px;
}
blockquote {
margin: 0;
}
ul,
ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
ol ol,
ul ol {
list-style-type: lower-roman;
}
ul ul ol,
ul ol ol,
ol ul ol,
ol ol ol {
list-style-type: lower-alpha;
}
dd {
margin-left: 0;
}
code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
.select::-ms-expand {
opacity: 0;
}
.octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
.octicon-link:before {
content: '\f05c';
}
.markdown-body:before {
display: table;
content: "";
}
.markdown-body:after {
display: table;
clear: both;
content: "";
}
.markdown-body>*:first-child {
margin-top: 0 !important;
}
.markdown-body>*:last-child {
margin-bottom: 0 !important;
}
a:not([href]) {
color: inherit;
text-decoration: none;
}
.anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
.anchor:focus {
outline: none;
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
h1 .octicon-link,
h2 .octicon-link,
h3 .octicon-link,
h4 .octicon-link,
h5 .octicon-link,
h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
h1:hover .anchor,
h2:hover .anchor,
h3:hover .anchor,
h4:hover .anchor,
h5:hover .anchor,
h6:hover .anchor {
text-decoration: none;
}
h1:hover .anchor .octicon-link,
h2:hover .anchor .octicon-link,
h3:hover .anchor .octicon-link,
h4:hover .anchor .octicon-link,
h5:hover .anchor .octicon-link,
h6:hover .anchor .octicon-link {
visibility: visible;
}
h1 {
padding-bottom: 0.3em;
font-size: 2.25em;
line-height: 1.2;
border-bottom: 1px solid #eee;
}
h1 .anchor {
line-height: 1;
}
h2 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.225;
border-bottom: 1px solid #eee;
}
h2 .anchor {
line-height: 1;
}
h3 {
font-size: 1.5em;
line-height: 1.43;
}
h3 .anchor {
line-height: 1.2;
}
h4 {
font-size: 1.25em;
}
h4 .anchor {
line-height: 1.2;
}
h5 {
font-size: 1em;
}
h5 .anchor {
line-height: 1.1;
}
h6 {
font-size: 1em;
color: #777;
}
h6 .anchor {
line-height: 1.1;
}
p,
blockquote,
ul,
ol,
dl,
table,
pre {
margin-top: 0;
margin-bottom: 16px;
}
hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
ul,
ol {
padding-left: 2em;
}
ul ul,
ul ol,
ol ol,
ol ul {
margin-top: 0;
margin-bottom: 0;
}
li>p {
margin-top: 16px;
}
dl {
padding: 0;
}
dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
blockquote>:first-child {
margin-top: 0;
}
blockquote>:last-child {
margin-bottom: 0;
}
table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
table th {
font-weight: bold;
}
table th,
table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
table tr:nth-child(2n) {
background-color: #f8f8f8;
}
img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
code:before,
code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
.highlight {
margin-bottom: 16px;
}
.highlight pre,
pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
.highlight pre {
margin-bottom: 0;
word-break: normal;
}
pre {
word-wrap: normal;
}
pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
pre code:before,
pre code:after {
content: normal;
}
kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.pl-c {
color: #969896;
}
.pl-c1,
.pl-s .pl-v {
color: #0086b3;
}
.pl-e,
.pl-en {
color: #795da3;
}
.pl-s .pl-s1,
.pl-smi {
color: #333;
}
.pl-ent {
color: #63a35c;
}
.pl-k {
color: #a71d5d;
}
.pl-pds,
.pl-s,
.pl-s .pl-pse .pl-s1,
.pl-sr,
.pl-sr .pl-cce,
.pl-sr .pl-sra,
.pl-sr .pl-sre {
color: #183691;
}
.pl-v {
color: #ed6a43;
}
.pl-id {
color: #b52a1d;
}
.pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
.pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
.pl-ml {
color: #693a17;
}
.pl-mh,
.pl-mh .pl-en,
.pl-ms {
color: #1d3e81;
font-weight: bold;
}
.pl-mq {
color: #008080;
}
.pl-mi {
color: #333;
font-style: italic;
}
.pl-mb {
color: #333;
font-weight: bold;
}
.pl-md {
background-color: #ffecec;
color: #bd2c00;
}
.pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
.pl-mdr {
color: #795da3;
font-weight: bold;
}
.pl-mo {
color: #1d3e81;
}
kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.task-list-item {
list-style-type: none;
}
.task-list-item+.task-list-item {
margin-top: 3px;
}
.task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
:checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
.sourceLine {
display: inline-block;
}
code .kw { color: #000000; }
code .dt { color: #ed6a43; }
code .dv { color: #009999; }
code .bn { color: #009999; }
code .fl { color: #009999; }
code .ch { color: #009999; }
code .st { color: #183691; }
code .co { color: #969896; }
code .ot { color: #0086b3; }
code .al { color: #a61717; }
code .fu { color: #63a35c; }
code .er { color: #a61717; background-color: #e3d2d2; }
code .wa { color: #000000; }
code .cn { color: #008080; }
code .sc { color: #008080; }
code .vs { color: #183691; }
code .ss { color: #183691; }
code .im { color: #000000; }
code .va {color: #008080; }
code .cf { color: #000000; }
code .op { color: #000000; }
code .bu { color: #000000; }
code .ex { color: #000000; }
code .pp { color: #999999; }
code .at { color: #008080; }
code .do { color: #969896; }
code .an { color: #008080; }
code .cv { color: #008080; }
code .in { color: #008080; }
</style>
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="bitou-abc-analysis">Bitou ABC analysis</h1>
<p>This document outlines the different steps in the ABC random forest and outlier analyses carried out on bitou bush ddRAD SNPs in 2019.</p>
<h2 id="convert-vcf-to-diyabc-input-file">Convert VCF to DIYABC input file</h2>
<p>Starting with the full bitou SNP set, the first step in the ABC analysis is to prepare a SNP matrix as input for DIYABC. It is worth noting that the order of populations in the input SNP matrix, is the order that is used in the ABC analysis! So “sample 1” in the ABC parameters corresponds to the first population in the SNP input file.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># Enter working directory containing VCF file</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="bu">cd</span> <span class="st">&quot;/mnt/d/Documents/UWA/Applied Bioinformatics/Bitou/DIYABC_2019/final_input_snps&quot;</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co"># To reduce the size of the SNP set, we can remove all SNPs with any missing data</span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="ex">vcftools</span> --vcf bitou_core_snps_nopisi.vcf --max-missing 1 --recode --stdout <span class="op">&gt;</span> bitou_core_snps_nopisi_nomiss.vcf</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co"># We can then check how many SNPs are left</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co">#grep -v &#39;#&#39; bitou_core_snps_nopisi_nomiss.vcf | wc -l</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co"># Alternatively we can retain SNPs with up to 95% missing genotypes</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co">##vcftools --vcf bitou_core_snps_nopisi.vcf --max-missing 0.95 --recode --stdout &gt; bitou_core_snps_nopisi_mm95.vcf</span></span></code></pre></div>
<p>Next we should ensure that each population of the five defined in our popmap file is represented. Stacks populations is installed in a conda environment, so we need to load this first. We can skip this step since we have no missing data in our dataset.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># Enter working directory containing VCF file</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="bu">cd</span> <span class="st">&quot;/mnt/d/Documents/UWA/Applied Bioinformatics/Bitou/DIYABC_2019/final_input_snps&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="ex">conda</span> activate runstacks</span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="ex">populations</span> -V bitou_core_snps_nopisi_nomiss.vcf -O . -M final_popmap.txt -p 5 --vcf --no_hap_exports</span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="ex">conda</span> deactivate</span></code></pre></div>
<p>Now we need to remove monomorphic SNPs. These should have been removed with the MAF filter earlier, however, this does not exclude monomorphic heterozygous genotypes. A python script can filter out any monomorphic SNPs.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1"></a><span class="im">import</span> sys</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="im">import</span> csv</span>
<span id="cb3-3"><a href="#cb3-3"></a></span>
<span id="cb3-4"><a href="#cb3-4"></a>vcf <span class="op">=</span> sys.argv[<span class="dv">1</span>] <span class="co">#input vcf file</span></span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="cf">with</span> <span class="bu">open</span>(vcf) <span class="im">as</span> tsv:</span>
<span id="cb3-7"><a href="#cb3-7"></a>    <span class="cf">for</span> line <span class="kw">in</span> csv.reader(tsv, dialect<span class="op">=</span><span class="st">&quot;excel-tab&quot;</span>):</span>
<span id="cb3-8"><a href="#cb3-8"></a>        <span class="co"># skip comment lines</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>        <span class="cf">if</span> <span class="kw">not</span> line[<span class="dv">0</span>].startswith(<span class="st">&quot;#&quot;</span>):</span>
<span id="cb3-10"><a href="#cb3-10"></a>            <span class="co"># number of non-genotype fields in standard VCF format</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>            num_info_fields <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>        <span class="co"># get number of samples</span></span>
<span id="cb3-13"><a href="#cb3-13"></a>            num_samples <span class="op">=</span> <span class="bu">len</span>(line) <span class="op">-</span> num_info_fields</span>
<span id="cb3-14"><a href="#cb3-14"></a>            het_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-15"><a href="#cb3-15"></a>            alt_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-16"><a href="#cb3-16"></a>            hom_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-17"><a href="#cb3-17"></a>            null_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-18"><a href="#cb3-18"></a>        <span class="co"># loop through alleles for each SNP and count genotypes</span></span>
<span id="cb3-19"><a href="#cb3-19"></a>            <span class="cf">for</span> allele <span class="kw">in</span> line[num_info_fields:]:</span>
<span id="cb3-20"><a href="#cb3-20"></a>                g <span class="op">=</span> allele.split(<span class="st">&quot;:&quot;</span>)[<span class="dv">0</span>]</span>
<span id="cb3-21"><a href="#cb3-21"></a>                <span class="cf">if</span> g <span class="op">==</span> <span class="st">&quot;0/1&quot;</span>:</span>
<span id="cb3-22"><a href="#cb3-22"></a>                    het_count <span class="op">=</span> het_count <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb3-23"><a href="#cb3-23"></a>                <span class="cf">elif</span> g <span class="op">==</span> <span class="st">&quot;1/1&quot;</span>:</span>
<span id="cb3-24"><a href="#cb3-24"></a>                    alt_count <span class="op">=</span> alt_count <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb3-25"><a href="#cb3-25"></a>                <span class="cf">elif</span> g <span class="op">==</span> <span class="st">&quot;0/0&quot;</span>:</span>
<span id="cb3-26"><a href="#cb3-26"></a>                    hom_count <span class="op">=</span> hom_count <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb3-27"><a href="#cb3-27"></a>                <span class="cf">elif</span> g <span class="op">==</span> <span class="st">&quot;./.&quot;</span>:</span>
<span id="cb3-28"><a href="#cb3-28"></a>                    null_count <span class="op">=</span> null_count <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb3-29"><a href="#cb3-29"></a></span>
<span id="cb3-30"><a href="#cb3-30"></a>            <span class="co"># print SNP if user-defined thresholds are met</span></span>
<span id="cb3-31"><a href="#cb3-31"></a>            tot <span class="op">=</span> null_count <span class="op">+</span> het_count</span>
<span id="cb3-32"><a href="#cb3-32"></a>            tot2 <span class="op">=</span> null_count <span class="op">+</span> hom_count</span>
<span id="cb3-33"><a href="#cb3-33"></a>            tot3 <span class="op">=</span> null_count  <span class="op">+</span> alt_count</span>
<span id="cb3-34"><a href="#cb3-34"></a>            <span class="cf">if</span> tot <span class="op">==</span> num_samples <span class="kw">or</span> tot2 <span class="op">==</span> num_samples <span class="kw">or</span> tot3 <span class="op">==</span> num_samples:</span>
<span id="cb3-35"><a href="#cb3-35"></a>                <span class="cf">pass</span></span>
<span id="cb3-36"><a href="#cb3-36"></a>            <span class="cf">else</span>:</span>
<span id="cb3-37"><a href="#cb3-37"></a>                <span class="bu">print</span>(<span class="op">*</span>line, sep <span class="op">=</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</span>
<span id="cb3-38"><a href="#cb3-38"></a>    <span class="co"># print all comment lines</span></span>
<span id="cb3-39"><a href="#cb3-39"></a>        <span class="cf">else</span>:</span>
<span id="cb3-40"><a href="#cb3-40"></a>            <span class="bu">print</span>(<span class="op">*</span>line, sep <span class="op">=</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</span></code></pre></div>
<p>We can use the script like so:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># Enter working directory containing VCF file</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="bu">cd</span> <span class="st">&quot;/mnt/d/Documents/UWA/Applied Bioinformatics/Bitou/DIYABC_2019/final_input_snps&quot;</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="ex">python3</span> filtmono.py bitou_core_snps_nopisi_nomiss.p.vcf <span class="op">&gt;</span> bitou_core_snps_nopisi_nomiss_nomono.p.vcf</span></code></pre></div>
<p>If there are now too many SNPs, we can downsample to a chosen number of SNPs, e.g. 4000</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># Enter working directory containing VCF file</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="bu">cd</span> <span class="st">&quot;/mnt/d/Documents/UWA/Applied\ Bioinformatics/Bitou/DIYABC_2019/final_input_snps&quot;</span></span>
<span id="cb5-3"><a href="#cb5-3"></a> <span class="kw">(</span><span class="fu">grep</span> <span class="st">&#39;#&#39;</span> bitou_core_snps_nopisi_nomiss_nomono.p.vcf <span class="kw">&amp;&amp;</span> <span class="fu">grep</span> -v <span class="st">&#39;#&#39;</span>bitou_core_snps_nopisi_nomiss_nomono.p.vcf <span class="kw">|</span> <span class="ex">shuf</span> -n 4000<span class="kw">)</span> <span class="op">&gt;</span><span class="ex">bitou_core_snps_nopisi_nomiss_nomono_4k.p.vcf</span></span></code></pre></div>
<p>Finally we can convert to DIYABC SNP format. I have modified the vcf2diyabc.py (<a href="https://github.com/loire/vcf2DIYABC.py">https://github.com/loire/vcf2DIYABC.py</a>) script to directly take user arguments.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># Enter working directory containing VCF file</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="bu">cd</span> <span class="st">&quot;/mnt/d/Documents/UWA/Applied Bioinformatics/Bitou/DIYABC_2019/final_input_snps&quot;</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co"># To convert to DIYABC SNP format we need a popmap file with a dummy sex column, for example:</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="bu">echo</span> <span class="st">&quot;DIYABC popmap file looks like this:&quot;</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="fu">head</span> final_pop_diyabc_all.txt</span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="bu">printf</span> <span class="st">&quot;\n\n&quot;</span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co"># We can now conver to SNP format using vcf2diyabc.py</span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="ex">python2</span> vcf2DIYABC_clean.py bitou_core_snps_nopisi_nomiss_nomono_4k.p.vcf final_pop_diyabc_all.txt</span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co"># The resulting SNP file can then be sorted by population</span></span>
<span id="cb6-10"><a href="#cb6-10"></a> <span class="kw">(</span><span class="fu">head</span> -n 3 bitou_core_snps_nopisi_nomiss_nomono_4k.p.DIYABC.snp <span class="kw">&amp;&amp;</span> <span class="fu">tail</span> -n +4 bitou_core_snps_nopisi_nomiss_nomono_4k.p.DIYABC.snp<span class="kw">|</span> <span class="fu">sort</span><span class="kw">)</span> <span class="op">&gt;</span> <span class="ex">bitou_core_snps_nopisi_nomiss_nomono_4k.p.DIYABC.sort.snp</span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co"># We shouldn&#39;t forget to add the sex ratio to the SNP, as this leads to a loading error in DIYABC</span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="fu">sed</span> -i <span class="st">&#39;1s/insert title here &lt;NM=$insert_sex_ratio$NF&gt;/bitou &lt;NM=1.0NF&gt;/&#39;</span> bitou_core_snps_nopisi_nomiss_nomono_4k.p.DIYABC.sort.snp</span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co"># The sex ratio is used for non-autosomal variants in animals and will not be used with our SNPs</span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co"># We therefore use a dummy value of 1.0</span></span></code></pre></div>
<pre><code>## DIYABC popmap file looks like this:
## EAU_NSW_WOL_BB_02    9   EAU
## EAU_NSW_WOL_BB_03    9   EAU
## EAU_NSW_DUN_BB_02    9   EAU
## EAU_NSW_DUN_BB_03    9   EAU
## EAU_NSW_ILUKA_BB_02  9   EAU
## EAU_NSW_ILUKA_BB_03  9   EAU
## EAU_NSW_MINNIE_BB_01 9   EAU
## EAU_NSW_MINNIE_BB_02 9   EAU
## EAU_NSW_WOL_BB_04    9   EAU
## EAU_NSW_DUN_BB_01    9   EAU
## 
## 
## Parsing bitou_core_snps_nopisi_nomiss_nomono_4k.p.vcf...
## Writing outputfile as: bitou_core_snps_nopisi_nomiss_nomono_4k.p.DIYABC.snp
</code></pre>
<h2 id="diyabc-parameters---using-header-files">DIYABC parameters - using header files</h2>
<p>The header file specifies all model parameters and scenarios to be tested. Together with the SNP input file, it makes up the core input for DIYABC analysis. There are two data dryad repositories for published data that provide excellent examples of a range of SNP input data and ABC paramter header files: <a href="https://datadryad.org/resource/doi:10.5061/dryad.f6154">Extraordinarily rapid speciation in a marine fish</a> and <a href="https://datadryad.org/resource/doi:10.5061/dryad.88589">Molecular data reveal recent genetic diversification in a hummingbird</a>. These header files can be generated via text editor, or automatically using the DIYABC user interface.</p>
<p>Below is an example of a scenario that was tested in bitou. Note that in the first line we have parameters for the population size of 7 populations, although we only sampled 5 populations. The populations 6 (RSA_NA) and 7 (RSA) represent an unsampled South African population and an unsampled South African ancestral population respectively. The ancestral population is required because we do not know how the four other South African populations are related.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">EAU</span> DUR EBEACH STJOHN WAU RSA_NA RSA</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="ex">0</span> sample 1 <span class="co"># population 1 (first occuring in SNP input file) sampled at time 0</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="ex">0</span> sample 2 <span class="co"># population 2 (second occuring in SNP input file) sampled at time 0</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="ex">0</span> sample 3 <span class="co"># population 3 (third occuring in SNP input file) sampled at time 0</span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="ex">0</span> sample 4 <span class="co"># population 4 (fourth occuring in SNP input file) sampled at time 0</span></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="ex">0</span> sample 5 <span class="co"># population 5 (fifth occuring in SNP input file) sampled at time 0</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="ex">t_invW-db2</span> VarNe 5 WAU0 <span class="co"># bottleneck after colonization of WAU</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="ex">t_adm</span> split 5 6 1 ra <span class="co"># admixture between EAU population and South African ghost population</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="ex">t_invE-db1</span> VarNe 1 EAU0 <span class="co"># bottleneck after colonization of EAU</span></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="ex">t_invE</span> merge 6 1 <span class="co"># South African ghost population and EAU population diverge at the time of invasion t_invE</span></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="ex">t_anc</span> merge 7 2</span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="ex">t_anc</span> merge 7 3</span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="ex">t_anc</span> merge 7 4</span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="ex">t_anc</span> merge 7 6</span>
<span id="cb8-15"><a href="#cb8-15"></a></span>
<span id="cb8-16"><a href="#cb8-16"></a><span class="co"># three priors for effective population size of four populations</span></span>
<span id="cb8-17"><a href="#cb8-17"></a><span class="ex">0</span> sample 1 </span>
<span id="cb8-18"><a href="#cb8-18"></a><span class="ex">0</span> sample 2</span>
<span id="cb8-19"><a href="#cb8-19"></a><span class="ex">0</span> sample 3 </span></code></pre></div>
<p>The priors set for the above scenario are as follows.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># Below are the population sizes uniform distributions</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="ex">EAU</span> N UN[10.0,10000.0,0.0,0.0] <span class="co"># Uniform distribution of 10 diploid individuals to 10k diploid individuals</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="ex">DUR</span> N UN[10.0,10000.0,0.0,0.0]</span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="ex">EBEACH</span> N UN[10.0,10000.0,0.0,0.0]</span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="ex">STJOHN</span> N UN[10.0,10000.0,0.0,0.0]</span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="ex">WAU</span> N UN[10.0,10000.0,0.0,0.0]</span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="ex">RSA_NA</span> N UN[10.0,10000.0,0.0,0.0]</span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="ex">RSA</span> N UN[10.0,10000.0,0.0,0.0]</span>
<span id="cb9-9"><a href="#cb9-9"></a></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="co"># Below here are the time parameters (in generations) and additional bottleneck population sizes</span></span>
<span id="cb9-11"><a href="#cb9-11"></a><span class="ex">t_invW</span> T UN[8.0,10.0,0.0,0.0] <span class="co"># time of invasion of Western Australia in generations</span></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="ex">db2</span> T UN[1.0,5.0,0.0,0.0] <span class="co"># WAU bottleneck duration is fixed to 1-5 generations</span></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="ex">WAU0</span> N UN[5.0,1000.0,0.0,0.0] <span class="co"># WAU bottleneck population size</span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="ex">t_adm</span> T UN[9.0,44.0,0.0,0.0] <span class="co"># time of admixture between EAU and RSA_NA pop</span></span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="ex">ra</span> A UN[0.001,0.5,0.0,0.0] <span class="co"># admixture rate is bounded to maximum 0.5, with 1.0 being maximum admixture</span></span>
<span id="cb9-16"><a href="#cb9-16"></a><span class="ex">t_invE</span> T UN[38.0,45.0,0.0,0.0] <span class="co"># time of invasion of Eastern Australia in generations</span></span>
<span id="cb9-17"><a href="#cb9-17"></a><span class="ex">db1</span> T UN[1.0,5.0,0.0,0.0] <span class="co"># EAU bottleneck duration is fixed to 1-5 generations</span></span>
<span id="cb9-18"><a href="#cb9-18"></a><span class="ex">EAU0</span> N UN[5.0,1000.0,0.0,0.0] <span class="co"># EAU bottleneck population size</span></span>
<span id="cb9-19"><a href="#cb9-19"></a><span class="ex">t_anc</span> T UN[10.0,10000.0,0.0,0.0] <span class="co"># The time all the populations split from the ancestral population</span></span>
<span id="cb9-20"><a href="#cb9-20"></a></span>
<span id="cb9-21"><a href="#cb9-21"></a><span class="co"># Time constraints</span></span>
<span id="cb9-22"><a href="#cb9-22"></a><span class="ex">t_anc</span><span class="op">&gt;</span>t_invW</span>
<span id="cb9-23"><a href="#cb9-23"></a><span class="ex">t_anc</span><span class="op">&gt;</span>t_invE</span>
<span id="cb9-24"><a href="#cb9-24"></a><span class="ex">t_invE</span><span class="op">&gt;</span>t_invW</span>
<span id="cb9-25"><a href="#cb9-25"></a><span class="ex">t_anc</span><span class="op">&gt;</span>t_adm</span>
<span id="cb9-26"><a href="#cb9-26"></a><span class="ex">t_invE</span><span class="op">&gt;</span>t_adm</span>
<span id="cb9-27"><a href="#cb9-27"></a><span class="ex">t_adm</span><span class="op">&gt;</span>t_invW</span></code></pre></div>
<h2 id="running-diyabc-on-a-cluster">Running DIYABC on a cluster</h2>
<p>Based on my experience, I do not recommend running DIYABC on a Linux cluster. Many analyses resulted in segmentation faults, which are hard to trace, e.g. for model checking and parameter estimation. Additionally, I was not able to import the reftables generated via the command line into the DIYABC GUI program to carry out these analyses.</p>
<p>To run the analysis on a cluster, the DIYABC linux binary is required, which is easy to compile. In the GUI, under ‘settings’ it is possible to select cluster, which when selected will generate an archive of useful files when the user hits the main ‘analyse’ button to start the simulations. The key files are the node.sh and the launch.sh files. The launch.sh can be used to generate RNG files required for the parallelized analysis. The node.sh script can launch the analysis. These scripts did not work out of the box on my server, so I modified them. The launch script should be modified to generate the desired number of RNG files (eg 200 for 200 parallel jobs). These files should then be moved into separate directores named “mtymp_$RNGNUM”. I modified the node.sh script to use this name for temporary working directories in which reftables are then generated. The node.sh script is then launched in parallel for each RNG file.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="va">DIYABCEXE=</span><span class="st">&quot;/scratch/pawsey0149/ascheben/bitou/diyabc3/diyabc_core-2.1.0-linux-x64&quot;</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="va">CPU=</span>1</span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co"># Random number generator file</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="va">RNGNUM=</span>1</span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co"># Number of generations to run simulations</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="va">NGEN=</span>1000000</span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co"># Working dir</span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="va">WD=</span><span class="st">&quot;/scratch/pawsey0149/ascheben/bitou/diyabc3&quot;</span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co"># Input SNP file</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="va">INSNP=</span><span class="st">&quot;/scratch/pawsey0149/ascheben/bitou/diyabc3/bitou_core_snps_nopisi_nomiss.p.snps.DIYABC.snp&quot;</span></span>
<span id="cb10-11"><a href="#cb10-11"></a></span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="fu">bash</span> node.sh <span class="st">&#39;$DIYABCEXE&#39;</span> <span class="va">$CPU</span> <span class="va">$NGEN</span> <span class="st">&#39;$WD&#39;</span> <span class="va">$RNGNUM</span> <span class="st">&#39;$INSNP&#39;</span></span></code></pre></div>
<p>This is the modified node.sh script.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">set</span> <span class="ex">-o</span> nounset</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="bu">echo</span> <span class="st">&quot;host =  </span><span class="kw">`</span><span class="fu">hostname</span><span class="kw">`</span><span class="st">&quot;</span></span>
<span id="cb11-3"><a href="#cb11-3"></a></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co"># ===============================================</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co"># check script parameters</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="kw">function</span><span class="fu"> usage</span> <span class="kw">{</span></span>
<span id="cb11-7"><a href="#cb11-7"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\tThis script runs DIYABC on a node cluster&quot;</span></span>
<span id="cb11-8"><a href="#cb11-8"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\tUsage : </span><span class="va">$0</span><span class="st"> &lt;diyabcPath&gt; &lt;numberOfCores&gt; &lt;numToGenerate&gt; &lt;workingDirectory&gt; &lt;numIdOfTheCurrentJob&gt; &lt;dataFile&gt;&quot;</span></span>
<span id="cb11-9"><a href="#cb11-9"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\texample : &quot;</span></span>
<span id="cb11-10"><a href="#cb11-10"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\t\t </span><span class="va">$0</span><span class="st"> </span><span class="dt">\&quot;</span><span class="st">/usr/local/bin/diyabc</span><span class="dt">\&quot;</span><span class="st"> 1 5000 </span><span class="dt">\&quot;</span><span class="va">$PWD</span><span class="dt">\&quot;</span><span class="st"> 13 </span><span class="dt">\&quot;</span><span class="st">dyabcData.mss</span><span class="dt">\&quot;</span><span class="st">&quot;</span></span>
<span id="cb11-11"><a href="#cb11-11"></a><span class="kw">}</span></span>
<span id="cb11-12"><a href="#cb11-12"></a></span>
<span id="cb11-13"><a href="#cb11-13"></a><span class="co"># Test  parameters number</span></span>
<span id="cb11-14"><a href="#cb11-14"></a><span class="kw">if</span><span class="bu"> [</span> <span class="st">&quot;</span><span class="va">$#</span><span class="st">&quot;</span> <span class="ot">-lt</span> 5<span class="bu"> ]</span></span>
<span id="cb11-15"><a href="#cb11-15"></a>    <span class="kw">then</span></span>
<span id="cb11-16"><a href="#cb11-16"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\nERROR :&quot;</span></span>
<span id="cb11-17"><a href="#cb11-17"></a>    <span class="bu">echo</span> -e <span class="st">&quot;      Require 6  parameters, </span><span class="va">$#</span><span class="st"> given\n\n&quot;</span></span>
<span id="cb11-18"><a href="#cb11-18"></a>    <span class="ex">usage</span></span>
<span id="cb11-19"><a href="#cb11-19"></a>    <span class="bu">exit</span> 2</span>
<span id="cb11-20"><a href="#cb11-20"></a><span class="kw">fi</span></span>
<span id="cb11-21"><a href="#cb11-21"></a></span>
<span id="cb11-22"><a href="#cb11-22"></a><span class="va">DIYABCPATH=$1</span></span>
<span id="cb11-23"><a href="#cb11-23"></a><span class="va">NBCORES=$2</span></span>
<span id="cb11-24"><a href="#cb11-24"></a><span class="va">NBTOGEN=$3</span></span>
<span id="cb11-25"><a href="#cb11-25"></a><span class="va">USERDIR=$4</span></span>
<span id="cb11-26"><a href="#cb11-26"></a><span class="va">MYNUMBER=$5</span></span>
<span id="cb11-27"><a href="#cb11-27"></a><span class="va">DATAFILE=$6</span></span>
<span id="cb11-28"><a href="#cb11-28"></a><span class="va">MYTMPDIR=</span><span class="st">&quot;/scratch/pawsey0149/ascheben/bitou/diyabc2/mytmp_</span><span class="va">$MYNUMBER</span><span class="st">&quot;</span></span>
<span id="cb11-29"><a href="#cb11-29"></a><span class="co">#mkdir $MYTMPDIR</span></span>
<span id="cb11-30"><a href="#cb11-30"></a><span class="co"># Get full path for diyabc, dir and file</span></span>
<span id="cb11-31"><a href="#cb11-31"></a><span class="va">DIYABCPATH=</span><span class="kw">`</span><span class="fu">readlink</span> -f <span class="st">&quot;</span><span class="va">$DIYABCPATH</span><span class="st">&quot;</span><span class="kw">`</span></span>
<span id="cb11-32"><a href="#cb11-32"></a><span class="va">USERDIR=</span><span class="kw">`</span><span class="fu">readlink</span> -f <span class="st">&quot;</span><span class="va">$USERDIR</span><span class="st">&quot;</span><span class="kw">`</span></span>
<span id="cb11-33"><a href="#cb11-33"></a><span class="va">DATAFILE=</span><span class="kw">`</span><span class="fu">readlink</span> -f <span class="st">&quot;</span><span class="va">$DATAFILE</span><span class="st">&quot;</span><span class="kw">`</span></span>
<span id="cb11-34"><a href="#cb11-34"></a></span>
<span id="cb11-35"><a href="#cb11-35"></a><span class="kw">if</span><span class="bu"> [</span> <span class="ot">!</span> <span class="ot">-r</span> <span class="st">&quot;</span><span class="va">$DIYABCPATH</span><span class="st">&quot;</span><span class="bu"> ]</span>; <span class="kw">then</span></span>
<span id="cb11-36"><a href="#cb11-36"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\nERROR :&quot;</span> <span class="op">&gt;&amp;2</span></span>
<span id="cb11-37"><a href="#cb11-37"></a>    <span class="bu">echo</span> -e <span class="st">&quot;      diyabc not found in </span><span class="va">$DIYABCPATH</span><span class="st"> or you can not read it\n&quot;</span> <span class="op">&gt;&amp;2</span></span>
<span id="cb11-38"><a href="#cb11-38"></a>    <span class="ex">usage</span></span>
<span id="cb11-39"><a href="#cb11-39"></a>    <span class="bu">exit</span> 3</span>
<span id="cb11-40"><a href="#cb11-40"></a><span class="kw">fi</span></span>
<span id="cb11-41"><a href="#cb11-41"></a></span>
<span id="cb11-42"><a href="#cb11-42"></a><span class="kw">if</span><span class="bu"> [</span> <span class="ot">!</span> <span class="ot">-r</span> <span class="st">&quot;</span><span class="va">$DATAFILE</span><span class="st">&quot;</span><span class="bu"> ]</span>; <span class="kw">then</span></span>
<span id="cb11-43"><a href="#cb11-43"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\nERROR :&quot;</span> <span class="op">&gt;&amp;2</span></span>
<span id="cb11-44"><a href="#cb11-44"></a>    <span class="bu">echo</span> -e <span class="st">&quot;      You do not have rights to read </span><span class="va">$DATAFILE</span><span class="st">  but you should !\n&quot;</span> <span class="op">&gt;&amp;2</span></span>
<span id="cb11-45"><a href="#cb11-45"></a>    <span class="ex">usage</span></span>
<span id="cb11-46"><a href="#cb11-46"></a>    <span class="bu">exit</span> 3</span>
<span id="cb11-47"><a href="#cb11-47"></a><span class="kw">fi</span></span>
<span id="cb11-48"><a href="#cb11-48"></a></span>
<span id="cb11-49"><a href="#cb11-49"></a><span class="kw">if</span><span class="bu"> [</span> <span class="ot">!</span> <span class="ot">-w</span> <span class="st">&quot;</span><span class="va">$USERDIR</span><span class="st">&quot;</span><span class="bu"> ]</span>; <span class="kw">then</span></span>
<span id="cb11-50"><a href="#cb11-50"></a>    <span class="bu">echo</span> -e <span class="st">&quot;\nERROR :&quot;</span>  <span class="op">&gt;&amp;2</span></span>
<span id="cb11-51"><a href="#cb11-51"></a>    <span class="bu">echo</span> -e <span class="st">&quot;      You do not have the right to write in </span><span class="va">$USERDIR</span><span class="st"> but you should ! (or dir does)\n&quot;</span> <span class="op">&gt;&amp;2</span></span>
<span id="cb11-52"><a href="#cb11-52"></a>    <span class="ex">usage</span></span>
<span id="cb11-53"><a href="#cb11-53"></a>    <span class="bu">exit</span> 3</span>
<span id="cb11-54"><a href="#cb11-54"></a><span class="kw">fi</span></span>
<span id="cb11-55"><a href="#cb11-55"></a></span>
<span id="cb11-56"><a href="#cb11-56"></a><span class="bu">echo</span> -e <span class="st">&quot;All parameters accepted&quot;</span></span>
<span id="cb11-57"><a href="#cb11-57"></a><span class="co"># ===============================================</span></span>
<span id="cb11-58"><a href="#cb11-58"></a><span class="co"># start working</span></span>
<span id="cb11-59"><a href="#cb11-59"></a></span>
<span id="cb11-60"><a href="#cb11-60"></a><span class="va">jobId=</span><span class="st">&quot;</span><span class="kw">`</span><span class="fu">hostname</span><span class="kw">`</span><span class="st">-n</span><span class="va">${MYNUMBER}</span><span class="st">-pid</span><span class="va">${BASHPID}</span><span class="st">-</span><span class="va">${RANDOM}</span><span class="st">&quot;</span></span>
<span id="cb11-61"><a href="#cb11-61"></a></span>
<span id="cb11-62"><a href="#cb11-62"></a><span class="co"># initialisation values for trapActions</span></span>
<span id="cb11-63"><a href="#cb11-63"></a><span class="va">errStatus=</span><span class="st">&quot;&quot;</span></span>
<span id="cb11-64"><a href="#cb11-64"></a><span class="va">rngFile=</span><span class="st">&quot;&quot;</span></span>
<span id="cb11-65"><a href="#cb11-65"></a><span class="va">rngFileName=</span><span class="st">&quot;&quot;</span></span>
<span id="cb11-66"><a href="#cb11-66"></a><span class="va">rngLockFile=</span><span class="st">&quot;&quot;</span></span>
<span id="cb11-67"><a href="#cb11-67"></a><span class="va">rngFlagFile=</span><span class="st">&quot;&quot;</span></span>
<span id="cb11-68"><a href="#cb11-68"></a><span class="va">rngBackupFile=</span><span class="st">&quot;&quot;</span></span>
<span id="cb11-69"><a href="#cb11-69"></a><span class="va">scriptMadeTMPDIR=</span><span class="st">&quot;false&quot;</span></span>
<span id="cb11-70"><a href="#cb11-70"></a></span>
<span id="cb11-71"><a href="#cb11-71"></a><span class="co">#copy files to tmp dir</span></span>
<span id="cb11-72"><a href="#cb11-72"></a><span class="fu">cp</span> <span class="st">&quot;</span><span class="va">$DIYABCPATH</span><span class="st">&quot;</span> <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/&quot;</span></span>
<span id="cb11-73"><a href="#cb11-73"></a><span class="fu">chmod</span> +x <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$DIYABCPATH</span><span class="kw">`</span><span class="st">&quot;</span></span>
<span id="cb11-74"><a href="#cb11-74"></a><span class="fu">cp</span> header.txt <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/&quot;</span></span>
<span id="cb11-75"><a href="#cb11-75"></a><span class="fu">cp</span> <span class="st">&quot;</span><span class="va">$DATAFILE</span><span class="st">&quot;</span> <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/&quot;</span></span>
<span id="cb11-76"><a href="#cb11-76"></a></span>
<span id="cb11-77"><a href="#cb11-77"></a><span class="co"># Computations</span></span>
<span id="cb11-78"><a href="#cb11-78"></a><span class="co">##rngNum=`echo ${rngFileName:10} | cut -d&#39;.&#39; -f1`</span></span>
<span id="cb11-79"><a href="#cb11-79"></a></span>
<span id="cb11-80"><a href="#cb11-80"></a><span class="va">rngNum=</span><span class="kw">`</span><span class="fu">expr</span> <span class="va">$5</span> - 1<span class="kw">`</span></span>
<span id="cb11-81"><a href="#cb11-81"></a><span class="va">cmd=</span><span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$DIYABCPATH</span><span class="kw">`</span><span class="st"> -t </span><span class="va">$NBCORES</span><span class="st"> -p </span><span class="va">$MYTMPDIR</span><span class="st">/ -w </span><span class="va">$rngNum</span><span class="st"> -r </span><span class="va">$NBTOGEN</span><span class="st"> &quot;</span></span>
<span id="cb11-82"><a href="#cb11-82"></a><span class="bu">echo</span> -e <span class="st">&quot;Running computations :&quot;</span></span>
<span id="cb11-83"><a href="#cb11-83"></a><span class="bu">echo</span> -e <span class="st">&quot;</span><span class="va">$cmd</span><span class="st">&quot;</span></span>
<span id="cb11-84"><a href="#cb11-84"></a><span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$DIYABCPATH</span><span class="kw">`</span><span class="st">&quot;</span> <span class="ex">-t</span> <span class="va">$NBCORES</span> -p <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/&quot;</span> -w <span class="va">$rngNum</span> -r <span class="va">$NBTOGEN</span> <span class="op">&amp;&gt;</span> <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">/diyabc.out&quot;</span></span>
<span id="cb11-85"><a href="#cb11-85"></a><span class="co">#copy results to USERDIR and clean</span></span>
<span id="cb11-86"><a href="#cb11-86"></a><span class="bu">echo</span> -e <span class="st">&quot;computation finished with succes, bringing back results&quot;</span></span>
<span id="cb11-87"><a href="#cb11-87"></a><span class="fu">mv</span> <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">&quot;</span>/reftable.bin <span class="st">&quot;</span><span class="va">$USERDIR</span><span class="st">&quot;</span>/reftable_<span class="va">$MYNUMBER</span>.bin</span>
<span id="cb11-88"><a href="#cb11-88"></a><span class="fu">mv</span> <span class="st">&quot;</span><span class="va">$MYTMPDIR</span><span class="st">&quot;</span>/reftable.log <span class="st">&quot;</span><span class="va">$USERDIR</span><span class="st">&quot;</span>/reftable_<span class="va">$MYNUMBER</span>.log</span>
<span id="cb11-89"><a href="#cb11-89"></a><span class="co">##mv &quot;$MYTMPDIR/$rngFileName&quot; &quot;$rngFile&quot;</span></span>
<span id="cb11-90"><a href="#cb11-90"></a><span class="bu">echo</span> -e <span class="st">&quot;cleaning temp files&quot;</span></span>
<span id="cb11-91"><a href="#cb11-91"></a><span class="co">#rm -f &quot;$rngBackupFile&quot; &quot;$rngLockFile&quot; &quot;$rngFlagFile&quot;</span></span>
<span id="cb11-92"><a href="#cb11-92"></a><span class="co">##if [ $scriptMadeTMPDIR = &#39;true&#39; ]; then rm -rf &quot;$MYTMPDIR&quot;; fi</span></span>
<span id="cb11-93"><a href="#cb11-93"></a><span class="bu">echo</span> -e <span class="st">&quot;End with succes&quot;</span></span>
<span id="cb11-94"><a href="#cb11-94"></a><span class="bu">echo</span> -e <span class="st">&quot; &quot;</span></span>
<span id="cb11-95"><a href="#cb11-95"></a><span class="co">#</span><span class="re">END</span></span>
<span id="cb11-96"><a href="#cb11-96"></a><span class="bu">exit</span> 0</span></code></pre></div>
<p>All the completed reftable_${RNG}.bin files will be moved to the working directory. They can be concatenated using the DIYABC binary as follows. This will generate a final “reftable.bin” file.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1"></a><span class="va">$diyabcPath</span> <span class="ex">-p</span> <span class="st">&quot;</span><span class="va">$PWD</span><span class="st">/&quot;</span> -q  <span class="op">&amp;&gt;</span> concat.out</span></code></pre></div>
<h2 id="abc-random-forest-analysis">ABC Random forest analysis</h2>
<p>Based on the literature, DIYABC simulations are commonly run for 1 million or more generations per scenario. However, with a dataset of &gt;1k SNPs and a large number of scenarios to test, this is not feasible. ABC random forest analysis enables scenario selection with accuracies as high as DIYABC but using only ~10k generations of simulations. The R package abcrf can be used to implement this method. Three input files generated by DIYABC are required:</p>
<ol>
<li>reftable.bin</li>
<li>statobs.txt</li>
<li>header.txt</li>
</ol>
<p>These files include the simulations, the scenarios and the observed summary statistics. To understand the below it is important to note that DIYABC will not generate absolutely exactly equal numbers of simulations per scenario. For example when requesting 30k simulations for 3 scenatios, DIYABC may generate 9998 for one scenario, 1002 for another and 1000 for the third. Therefore one should also generate slightly more scenarios and then extract an exact number during the data preparation stage.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># Load libraries</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="kw">library</span>(abcrf)</span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="kw">library</span>(data.table)</span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="co"># Set working directory</span></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="kw">setwd</span>(<span class="st">&quot;D://Documents//UWA//Applied Bioinformatics//Bitou//DIYABC_2019//randomforest&quot;</span>)</span>
<span id="cb13-6"><a href="#cb13-6"></a></span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="co"># Set total number of generations (usually 10k per scenario in the reftable)</span></span>
<span id="cb13-8"><a href="#cb13-8"></a>i &lt;-<span class="st"> </span><span class="dv">60000</span></span>
<span id="cb13-9"><a href="#cb13-9"></a></span>
<span id="cb13-10"><a href="#cb13-10"></a><span class="co"># Set names of output files</span></span>
<span id="cb13-11"><a href="#cb13-11"></a>outerr &lt;-<span class="st"> </span><span class="kw">paste</span>(i,<span class="st">&quot;_bagerr_1K_WAU_sort_best_cor_2db.txt&quot;</span>)</span>
<span id="cb13-12"><a href="#cb13-12"></a>outpred &lt;-<span class="st"> </span><span class="kw">paste</span>(i,<span class="st">&quot;_prediction_1K_WAU_sort_best_cor_2db.txt&quot;</span>)</span>
<span id="cb13-13"><a href="#cb13-13"></a>outobject &lt;-<span class="st"> </span><span class="kw">paste</span>(i,<span class="st">&quot;_objects_1K_WAU_sort_best_cor_2db.RData&quot;</span>)</span>
<span id="cb13-14"><a href="#cb13-14"></a>outplot &lt;-<span class="st"> </span><span class="kw">paste</span> (i, <span class="st">&quot;_errplot_1K_WAU_sort_best_cor_2db.pdf&quot;</span>)</span>
<span id="cb13-15"><a href="#cb13-15"></a></span>
<span id="cb13-16"><a href="#cb13-16"></a><span class="co"># Load DIYABC reftable binary and corresponding header file</span></span>
<span id="cb13-17"><a href="#cb13-17"></a>myref &lt;-<span class="st"> </span><span class="kw">readRefTable</span>(<span class="dt">filename =</span> <span class="st">&quot;reftable_1K_WAU_sort_best_cor_2db.bin&quot;</span>, <span class="dt">header =</span> <span class="st">&quot;header_1K_WAU_sort_best_cor_2db.txt&quot;</span>, <span class="dt">N =</span> <span class="dv">62000</span>)</span>
<span id="cb13-18"><a href="#cb13-18"></a><span class="co"># Extract scenario indices</span></span>
<span id="cb13-19"><a href="#cb13-19"></a>modindex &lt;-<span class="st"> </span>myref<span class="op">$</span>scenarios</span>
<span id="cb13-20"><a href="#cb13-20"></a><span class="co"># Extract summary statistics</span></span>
<span id="cb13-21"><a href="#cb13-21"></a>sumsta &lt;-<span class="st"> </span>myref<span class="op">$</span>stats</span>
<span id="cb13-22"><a href="#cb13-22"></a><span class="co"># get n simulations for each modindex (=scenario)</span></span>
<span id="cb13-23"><a href="#cb13-23"></a>data1 &lt;-<span class="st"> </span><span class="kw">data.table</span>(modindex, sumsta)</span>
<span id="cb13-24"><a href="#cb13-24"></a>data1 &lt;-<span class="st"> </span>data1[, <span class="kw">head</span>(.SD, <span class="dv">10000</span>), by=modindex]</span>
<span id="cb13-25"><a href="#cb13-25"></a><span class="co"># Generate dataframe</span></span>
<span id="cb13-26"><a href="#cb13-26"></a>data1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(data1)</span>
<span id="cb13-27"><a href="#cb13-27"></a><span class="co"># Generate model with 1000 trees</span></span>
<span id="cb13-28"><a href="#cb13-28"></a><span class="co"># The below uses five CPU cores and reads in i total simulations from the reftable</span></span>
<span id="cb13-29"><a href="#cb13-29"></a>model.rf1 &lt;-<span class="st"> </span><span class="kw">abcrf</span>(modindex<span class="op">~</span>., <span class="dt">data =</span> data1, <span class="dt">ntree=</span><span class="dv">1000</span>, <span class="dt">paral=</span><span class="ot">TRUE</span>, <span class="dt">sampsize =</span> i, <span class="dt">ncores =</span> <span class="dv">5</span>)</span>
<span id="cb13-30"><a href="#cb13-30"></a><span class="co">#estimate error rate of model with ntrees and save plot</span></span>
<span id="cb13-31"><a href="#cb13-31"></a><span class="kw">pdf</span>(outplot) </span>
<span id="cb13-32"><a href="#cb13-32"></a>bagerr1 &lt;-<span class="st"> </span><span class="kw">err.abcrf</span>(model.rf1, data1)</span>
<span id="cb13-33"><a href="#cb13-33"></a><span class="kw">dev.off</span>()</span>
<span id="cb13-34"><a href="#cb13-34"></a><span class="kw">sink</span>(<span class="dt">file =</span> outerr)</span>
<span id="cb13-35"><a href="#cb13-35"></a><span class="kw">print</span>(bagerr1)</span>
<span id="cb13-36"><a href="#cb13-36"></a><span class="kw">sink</span>()</span>
<span id="cb13-37"><a href="#cb13-37"></a><span class="co"># Reading the observed dataset</span></span>
<span id="cb13-38"><a href="#cb13-38"></a>obs.poi &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;statobs_1K_WAU_sort_best_cor_2db.txt&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)</span>
<span id="cb13-39"><a href="#cb13-39"></a><span class="co"># Prediction based on observed statistics</span></span>
<span id="cb13-40"><a href="#cb13-40"></a>pred.obsPoi1 &lt;-<span class="st"> </span><span class="kw">predict</span>(<span class="dt">object =</span> model.rf1, <span class="dt">obs =</span> obs.poi,</span>
<span id="cb13-41"><a href="#cb13-41"></a>                        <span class="dt">training =</span> data1, <span class="dt">ntree=</span><span class="dv">1000</span>)</span>
<span id="cb13-42"><a href="#cb13-42"></a><span class="co"># Save prediction in file</span></span>
<span id="cb13-43"><a href="#cb13-43"></a><span class="kw">sink</span>(<span class="dt">file =</span> outpred)</span>
<span id="cb13-44"><a href="#cb13-44"></a><span class="kw">print</span>(<span class="kw">do.call</span>(rbind.data.frame, pred.obsPoi1))</span>
<span id="cb13-45"><a href="#cb13-45"></a><span class="kw">sink</span>()</span>
<span id="cb13-46"><a href="#cb13-46"></a><span class="co"># Save objects from analysis in R object for later use</span></span>
<span id="cb13-47"><a href="#cb13-47"></a><span class="kw">save</span>(model.rf1,pred.obsPoi1,bagerr1, <span class="dt">file =</span> outobject)</span></code></pre></div>
<p>We can observe that with 1000 trees in the random forest the prior error rate stablises.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">setwd</span>(<span class="st">&quot;D://Documents//UWA//Applied Bioinformatics//Bitou//DIYABC_2019//randomforest&quot;</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="kw">load</span>(<span class="st">&quot;70000 _objects_1k_WAU_sort_best.RData&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="kw">plot</span>(bagerr1)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAHgCAMAAABNUi8GAAAAV1BMVEUAAAAAADoAAGYAOjoAOpAAZrY6AAA6ADo6AGY6kNtmAABmADpmtrZmtv+QOgCQZgCQkGaQ2/+2ZgC2/7a2///bkDrb/7bb////tmb/25D//7b//9v////ZItwfAAAACXBIWXMAAA7DAAAOwwHHb6hkAAANGklEQVR4nO3di3baygGGUZza6TFt3IaGE3x5/+csQkAuDjKIGfFL2nutZrF6Eg3EX4RmkNDiDYItbv0EoItAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlWuFAF3CWWwVadnNMlUCJJlCiCZRoAiWaQIkmUKIlBXrBohdzERToovwojF5OoIsawzB2AiWaQImWE6hjUP4gKFCzeN5LChTeESjRBEo0gRJNoEQTKNEESjSBEk2gRBMo0QRKtIEDvfwLI5g3e1CiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiVYp0NVi8fD6tFgsvhTZHLNVJ9D1/fdtow9vby/LxwKbY76qBPr6tM1yc/f1rU312s0xY1UCfVlu39k3n769HX69bnPMmD0o0Sofg+5SvXpzzJdZPNGsgxJNoESrFOju7X3rxBxeoJyp0iRpsZ8bbRYmSVyj3jJTyzITV6m3UN+yUM9V7EGJVusYdL8LdQzKdSrN4l+W7Sz+xP5ToJzJOijRBEo0gRLNfZKIVmkd9EeI1kG5Rp096OvTyQ/h+2yO+ap2sshDyc0xW7WOQTcnT1XutTnmyiyeaAIlmkCJVjvQtWUmrmEPSjSBEk2gRHNVJ9Fc1Uk01yQRzVWdRLMHJZqrOonmqk6iWQclmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGhuQ0M0e1CiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiVan0Bflov776sTt+8oPS7z1iPQzd3X9f33l+VVhQqUs1weaHOXrub2XOuTN4otOi7zdnmgzX0Om0BP3eSw8LjMW/896OrkTbqOnj9/OfWfBMpZeh+DHu92+AeH+8x13DFeoJyl5yx+sbj72vGb97fotAflWpXWQV+WzQGAQLlWj0nSv9p95weTpNV2FytQrtU/0I+WmdaLR4FyrUsDXf2Y/ny0UP/8+R8C5Ur996Afe306PdMXKGdxsgjRegR6WOXsOgbd7jy7f49AOUuPQFf339cPXTP03QSpfbA5daQqUM7S57P4x7dNc7LI6Y86m09D9079LoFyln4nizz/9W33vxOa37J3arVUoJyl38kizUy+I1B7UErpcQzarNCvHrve4t+OZ5I4BuU6fZaZVg/NTL7zg6TDTP9kxALlLNZBidbvGHTAcZm3frP4Acdl3vqcUX/FxUhuQ8Nl+uxBuy/meHPJB+XUmSS9Pn20lxUoZ6k0i399eii5OWarf6DdZ9RvOi76vGhc5s06KNEESjQL9USrvVB/6khVoJxl4IX6y8dl3qos1Jccl3kzSSJatYV6V3VSQp9Anz9/9O12ruqkkD6TpF10Xd8PWvyaJGc/zVb/ddCOa5JKX9W5uPQPMBn910E7VpsK70EXF/8JJqPKHrTwVZ0CnbEqx6CFr+oU6IzVmcUXHPfwm/U5T+NYqDeLny1nMxHNZcdEczYT0ZzNRLRxTJKYLZMkopkkEc0kiWgmSUQzSSKaQInWJ9Dtm/z999V1c3mBcpY+k6S7r+v77839vIYYl3nrtw7anKz80f3iC43LvPVbB20CvW61SaCcpf8edNVxyUfBcZm33seg3Zd8lBuXees5ix/2kg/myzoo0QYO1H2SuIw9KNEESjSBEk2gRBMo0QRKNIESTaBEEyjRBEo0gRJNoEQTKNEESjSBEk2gRBMo0QRKNIESrVKg7hdPGXUCdb94CqkSaPH7xTNbVQItfb945sselGi1jkFL3i+eGas0iy96v3hmzDoo0QRKNAv1RBv/Qr1vypu00S8zLUpujDhjX6hfFN0acca+BxXoxI19oV6gEzf6hXrHoNM2/nVQs/hJG3+gTJpAiTZwoG5Dw2UqrYP+CNEJy1yjzh709emjWyELlLNUO1nkoeTmmK1ax6CbD26GLFDOYhZPNIESTaBEqx3o+obLTFZbJ2DCe1CnkUzBdAN1It4kCJRo072qU6CTMP6rOk9yDDoFY78mqYtZ/ASM/apOJm7Ke1AmYOxXdTJxo7+qsx+Hp2Mx3XXQj0ZX6CjMMlBLpOMhUKIJlGizDNQx6HjMM1Cz+NGYaaCMhUCJJlCiCZRoAiWaQIkm0N9ZgYoi0N9Yw88i0F91fgpq5zo8gf6qK1A71xsQ6K86AnWKyS0I9Dend5MCvQWB/u7kgaZAb0Gg53MMegMDBzru29CM9GmPmj0o0QRKNIESTaBEEyjRBEo0gRJNoEQTKNEESjSBluFT0EoEWoTzSGoRaAnOxKtGoCUItBqBliDQagRaRNcxqPnTNQRaxukKtXsVgVbW9zpm7bYEWlnP65h7rltNL2uBVtYv0J6zruLLsbcPvlKgq0/f3p4/LxZ3X4tsbsx6XWjfL9DOP9WntYDPH+oEuuvzn19/ue/xFZsbtz7zp/KB9mmt+1kMs3etEujL8nEb6UPz0N2Ou/Sb+5/e3E+/nv+fem2v8xmWTLdSoF8Ot+R2v/ieev2QC++SP9ojd/xjKPbzrfQWv917ru1Bh9fve3s6/9TFGyz7sVqdQF+Wn77tdqGbU7MkgfZXeOfa7826Z6AXP/day0yb9gtuHgptjh9KL5GWXtLqPVX74zO0Djo2Zd9Br9hgjwODHrMugY5NTKAdW+yzsz7x3yoF+vrUvsWfWAUVaH8Veiq8ve6RIgLdLdKvP33brdhfvzl+Vr6n4T7QvHxZoEqg+zXQ1XG5/rrN8avbf0DeX8dzH/AYdP8BZ7MGaqGecw03iz/sQR/a9/lrN8eM1TkGXTfr882B6MvSWzzXqDSLb9bp7zr6FCjnsQ5KNIESrfJC/allUIFynkqTpMVj+2BzeHDYyqjvk8Twai4zNZwPylVqLtQ3Ti7Uw1lqBHrGHnQ4A+6rhxtqki/quqEuPAbd70J/Pwa9gbH8BYeONJqhLvvDL8t2/3zz/ed4/oJDRxrNUKOd1YzlLzh0pNEMJdCkoSb5om4S6KmzmYYzlr/g0JFGM5Q9aNJQk3xRAp3OUJN8UQKdzlCTfFFDBvrhySJQVpmTRaCS0X7UyTwUPlkEyrIHJdpoTxZhHkZ7sgjzMNp1UOZBoEQTKNEESjSBEk2gRBtToLtTVdr7iB1uF7rpuG/otVa7xbTqQzX3Pn34ZYBqL2q92H97e+2hnv/69udhLh9vRIG+Pm1f27r5YTa3adrdqun4oILNbrW3+lCb7Si7bwus/6Kab8/cNIXWHqq5n9bbn4bpMd6IAn3+3PzrX+/vJdZ8j+7xQQUvyybQ6kO12x3kRb0+Pfw2QqWhNu35mO+H6TPeiAJtbf8BHks9Pqgwzvr+P9tAqw+1uy/F7kH1F3UMtPJQm8Xj7lyi98P0GW90ga62r3P3Q9389KD8MNtNN8eg1YfafPrfcndgPcCLOrzF1x+qDfTdMH3GG1ugzWkq7VHM9tfjg+LDNG9GTaDVh1o374bNvm2AF3WYotQfatfg+2H6jDeyQDeHOVLln2VzOuEwgd7tdyoDvKjm7lbPn4f4Bz7bQNvT/Oq/Re02PMhbfHtAtj04G+BFHQ8FvcVXsr8mqv58Yr3/jsAv1Ydqf1rbn1z9F3Xcfw0w1DwnSYfzpQdZZmr3oNWHai+j2QyxzNTmMchQm1kuMzWHT61BFurbT5KqD7U+/DOo/6IOx6ADfPowy4X6/ftu8/rWh0/M1tU/6qw+1Obw+W39F7Uaaqj9Yeb7YS4fb0SBMkcCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSbQajZfbv0MpkCgtbRfrMiVBFqLQIsQaGkvy38vd9/NvFjc//2v/zY3SVj/9LWHi8ePNsDPBFray7K5Id7d12YP+rJsvmO0uVHC7otjjw84m0BLe1k+7r5uuw30cf9/NN/penxw66c4JgItbXfsuf2lDfTL4eYF22SPD279FMdEoKW9D/Rwy5Djg1s/xTERaGkn9qBvle4eN3UCLe1doMf1JgtPPQi0tB+BPu6T3N3ucNXO7dsHnE2gpR33mqvF/d/tPrNZ/tzN3Y8POJdAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJZpAiSZQogmUaAIlmkCJJlCiCZRoAiWaQIkmUKIJlGgCJdr/AVOHXWkOecU8AAAAAElFTkSuQmCC" /><!-- --></p>
<p>We can check the outcome of the prediction. Here we can see as the header row the scenario numbers. The votes are how many of the trees supported each scenario. The posterior probability is only show here for the winning scenario (1), which is shown in the allocation row.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">print</span>(<span class="kw">do.call</span>(rbind.data.frame, pred.obsPoi1))</span></code></pre></div>
<pre><code>##                   1        2        3       4        5       6
## allocation   1.0000   1.0000   1.0000  1.0000   1.0000  1.0000
## vote       347.0000 205.0000 186.0000 75.0000 126.0000 61.0000
## post.prob    0.5169   0.5169   0.5169  0.5169   0.5169  0.5169
</code></pre>
<p>It is sometimes recommended to ensure that the scenario outcome remains the same with a different number of input simulations, so the analysis can be redone with 5,000 and 7,000 simulations per scenario. If the scenario selected remains the same, then 10k simulations are sufficient.</p>
<p>Finally, model checking and parameter estimation can be carried out for the best model using DIYABC. For example, the best model is rerun, this time with 100k-1m simulations in DIYABC. The DIYABC functions under ‘Analysis’ can then be used to see whether the posterior distribution clusters around the observed values as it should. It is sometimes suggested that model checking should make use of different summary statistics from model selection, and this can be done by not giving the whole table of summary statistics to abcrf and specifying the other statistics to DIYABC in the model checking analysis. The number of summary statistics significantly different from those in the posterior distribution of the best model can also be check by selecting ‘numerical results’ after the model check analysis. Many of these approaches are outlined in this <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/j.1365-294X.2010.04773.x">Molecular Ecology review</a> and in the <a href="http://www1.montpellier.inra.fr/CBGP/diyabc/">DIYABC manual</a>. The parameter estimation for the best scenario can be used to identify (with confidence intervals) the size of populations and the time of divergence or admixture events in the best model.</p>

</body>
</html>
